### ISSUE TITLE
SA1139 (Use literal suffix notation instead of casting) and unchecked

### ISSUE THREAD

**Issue by kindermannhubert** [DotNetAnalyzers/StyleCopAnalyzers#2275]
I think this analyzer should not report diagnostics inside unchedked blocks.

For example when you do something like this:
```c#
unchecked
{
    StandartMessageColor = (int)0xFFFFFFFF;
    ImportatntMessageColor = (int)0xFFFF0000;
}
```
resp.
```c#
var x = unchecked((ulong)-1);
```

you probably don't want to replace it with following:
```c#
unchecked
{
    StandartMessageColor = -1;
    ImportatntMessageColor = -65536;
}
```
resp.
```c#
var x = unchecked(18446744073709551615UL);
```

What do you think?

**Comment by vweijsters** [comment]
I agree with you that type casts in unchecked blocks should be exempt, as they can have relevant side effects.

**Comment by sharwell** [comment]
:thought_balloon: We could also restrict this analyzer to reporting cases where the literal value can be represented in the target type of the cast without any overflow or loss of precision. But then again that might be equivalent to disabling it in `unchecked` expressions.

**Comment by x3ntrix** [comment]
I have some flag enums which are used in native api calls. Therefore all values are defined as hex values like in the MSDN documentation. In some cases there is an unchecked cast which is reported by the analyzer, e.g.:
`Write_Through = unchecked((int)0x80000000)`  
In this case the readability is better with the hex value. If the analyzer is not disabled for all unchecked expressions, maybe it makes sense to disable it at least for flag enums.
I know that I could define the enum as uint and remove the unchecked cast which would prevent the stylecop warning, but this would trigger 'CA1028: Enum storage should be Int32'.
